### DISCUSSION TITLE
Multi-Column Unique blocks ForeignKey on MySQL

### DISCUSSION THREAD

**Discussion by d4fseeker** [sqlalchemy/sqlalchemy#7733] [Usage Questions] [answered]
## Describe the bug
SqlAlchemy-Migrate throws `"Foreign key constraint is incorrectly formed"`  when trying to use the same column for both a multi-column UNIQUE-Constraint and as a foreign-key.  According to MySQL documentation, the index is supposedly automatically added when a foreign-key is created, but that only works if I remove the Multicolumn Unique-Constraint.

I am lost on how I can force SqlAlchemy to workaround this issue by either creating the ForeignKey later or forcing the Index to be created explicitly?

### Code
```python
class Devices(db.Model):
  __tablename__ = "devices"
  id = db.Column(db.Integer, primary_key=True,autoincrement=True)
  system_id = db.Column(db.Integer, db.ForeignKey("systems.id"), index=True)
  system = db.relationship("Systems",foreign_keys=[system_id])
  path = db.Column(db.String(PATH_MAXLENGTH), index=True, nullable=False)
  #Make sure every device only exists once
  db.UniqueConstraint(system_id,path)

class Systems(db.Model):
  __tablename__ = "systems"
  id = db.Column(db.Integer, primary_key=True,autoincrement=True)
```

### SQL

Generated by the app:
```sql
CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(1024) NOT NULL,
        PRIMARY KEY (id),
        FOREIGN KEY(system_id) REFERENCES systems (id),
        CONSTRAINT unique_system_path UNIQUE (system_id, path)
)
```

Working (manually):
```sql
CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(1024) NOT NULL,
        PRIMARY KEY (id),
        INDEX(system_id),
        FOREIGN KEY(system_id) REFERENCES systems (id),
        CONSTRAINT unique_system_path UNIQUE (system_id, path)
)
```

## Versions
System:  Docker python:3.9-alpine
MySQL: MariaDB 10.6.5
SqlAlchemy:  1.4.31
DB Api: PyMysql 1.0.2
Flask-SqlAlchemy 2.5.1
Migrations using flask-migrate (Alembic)

**Comment by CaselIT** [comment]
Hi,

I probably don't understand the question fully, but looking at the wrong sql and right sql, can't you just add `index=True` to the `system_id` definition?

**Reply by d4fseeker** [reply]
Thanks, that was my first reaction too. But by mistake my example code above did not include that "index=True". I have updated the code now to clarify what I tried. 

Unfortunately the index=True seems to be ignored when a foreignkey is specified. Without knowing the inner workings of SqlAlchemy I can only guess that  due to the foreignkey already causing an index, this is optimized away?

**Comment by zzzeek** [comment]
are you using [sqlalchemy-migrate](https://pypi.org/project/sqlalchemy-migrate/) ?  that one?   that package is only used in some very legacy Openstack configurations and is long not-really-maintained, for many years, other than openstack's immediate needs, and I'm in the process of getting them to move completely off of that package in favor of Alembic migrations.  if that's what you're using, I wouldn't use it for any new development.

**Reply by d4fseeker** [reply]
Thank you. No I'm using Flask's module flask-migrate which uses Alembic. The module name is very confusing in this matter - sorry for using a keyword that means something entirely else in SqlAlchemy-speak than Flask.
```
Flask-Migrate is an extension that configures Alembic in the proper way to work with your Flask and Flask-SQLAlchemy application. 
In terms of the actual database migrations, everything is handled by Alembic so you get exactly the same functionality.
```

I am wiping the alembic folder and database between each try so Alembic only ever gets to do the initial setup which fails on this error.

**Comment by CaselIT** [comment]
I cannot reproduce. Please adapt the below script to reproduce the error
```py
import sqlalchemy as sa
from sqlalchemy.orm import declarative_base

Base = declarative_base()

class Devices(Base):
    __tablename__ = "devices"
    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
    system_id = sa.Column(sa.Integer, sa.ForeignKey("systems.id"))  # this will work with or without index=True.
    system = sa.orm.relationship("Systems", foreign_keys=[system_id])
    path = sa.Column(sa.String(256), index=True, nullable=False)
    # Make sure every device only exists once
    __table_args__ = (sa.UniqueConstraint(system_id, path),)

class Systems(Base):
    __tablename__ = "systems"
    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)

e = sa.create_engine(
    "mariadb+pymysql://scott:tiger@127.0.0.1:3306/test?charset=utf8mb4",
    echo=True,
)

Base.metadata.drop_all(e)
Base.metadata.create_all(e)
```

The genrated sql is 
```sql
CREATE TABLE systems (
        id INTEGER NOT NULL AUTO_INCREMENT,
        PRIMARY KEY (id)
)

CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(256) NOT NULL,
        PRIMARY KEY (id),
        UNIQUE (system_id, path),
        FOREIGN KEY(system_id) REFERENCES systems (id)
)

CREATE INDEX ix_devices_path ON devices (path)
```

**Reply by d4fseeker** [reply]
I think I found what goes wrong but I am now even more confused than at the beginning and would appreciate any enlightenment.

Your CREATE TABLE works on my DB _only_ as long as the Column `path` remains 256char. When changing it to 1024char, it no longer works. Basically this means that the error returned by Mysql is actually wrong as the issue is not linked to the foreignkey but the UNIQUE constraint which is limited to 256char (See this Django Bugreport that seems to be linked: https://code.djangoproject.com/ticket/28661 )

Your SQL works fine:
```
CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(256) NOT NULL,
        PRIMARY KEY (id),
        UNIQUE (system_id, path),
        FOREIGN KEY(system_id) REFERENCES systems (id)
)
```

Changing path to 1024 and it's KO
```
CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(1024) NOT NULL,
        PRIMARY KEY (id),
        UNIQUE (system_id, path),
        FOREIGN KEY(system_id) REFERENCES systems (id)
)
```

Adding INDEX(system_id) and it works again
```
CREATE TABLE devices (
        id INTEGER NOT NULL AUTO_INCREMENT,
        system_id INTEGER,
        path VARCHAR(1024) NOT NULL,
        PRIMARY KEY (id),
        UNIQUE (system_id, path),
        INDEX(system_id),
        FOREIGN KEY(system_id) REFERENCES systems (id)
)
```

I am by far no specialist on MySQL/MariaDB and even further from it in SqlAlchemy but it looks like a weird database engine behaviour to me unless I missed something completely. If I am able to force SqlAlchemy to add the index I should be fine but I don't see how I can do that. Alternatively I would have to add the UNIQUE later but that looks a bit hacky.

In case it is useful, the tables are currently InnoDB with utf8mb3_general_ci collation.

**Reply by CaselIT** [reply]
> looks like a weird database engine behaviour to me unless I missed something completely

That seems indeed the cause here.

I'm not sure if sqlalchemy supports inline index definition. @zzzeek do you remember if that's an option?

If you're adding the table with alembic you can modify the declaration so that the table is created first, than the index, then the unique constraint.

Alternatively, you may be able to replace both with a single unique index on system_id, path

**Reply by d4fseeker** [reply]
Thanks for the clarification. Based on your proposal I found a solution/workaround to my issue. While it seems impossible to force a index-creation or to delay the UNIQUE-Constraint, it *IS* possible to delay the ForeignKey Creation which results in the database accepting it.  By adding `use_alter=True` to the db.ForeignKey the creation works fine.
```python
system_id = db.Column(db.Integer, db.ForeignKey("systems.id", use_alter=True), index=True)
```

The reason for the whole issue apparently is that ForeignKey and Unique cash on the type of Index (BTREE vs HASH) due to the length of my `path` column. 

>Alternatively, you may be able to replace both with a single unique index on system_id, path

I'm not sure I am following you. I already have a single Unique-Constraint across both columns, distinct constraints would not have the desired effect unfortunately.

**Reply by CaselIT** [reply]
What I meant is that you could combine ` INDEX(system_id),` and `UNIQUE (system_id, path),` in a single unique index on `system_id, path`

**Reply by zzzeek** [reply]
I dont think we suppport that inline INDEX() syntax.  I've never seen that before and I'm not sure what it means.

**Reply by d4fseeker** [reply]
@zzzeek 
I'm somewhat confused regarding your statement now. I've taken the syntax from 1.4 "Defining Constraints and Indexes" documentation, 
>Indexes can be created anonymously (using an auto-generated name ix_<column label>) for a single column using the inline index keyword on [Column](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.Column)

@CaselIT I don't actually need the INDEX(system_id), but MySQL needs it for the foreignkey, otherwise it trips up on the UNIQUE-Constraint for some reason. This looks like a limitation in MySQL because when adding it via the ALTER TABLE command, it understands the need for an implicite Index-Creation just fine.

**Reply by zzzeek** [reply]
the "index" keyword in the API is inline but the DDL generated is:

```sql
CREATE TABLE my_table (... )

CREATE INDEX my_index ON my_table(some_column)
```

that's not the same DDL form as the "INDEX(col)" directive inline with the table definition.

**Comment by d4fseeker** [comment, marked as answer]
The solution to this issue is to delay the ForeignKey creation into an use_alter.  See for details:
https://github.com/sqlalchemy/sqlalchemy/discussions/7733#discussioncomment-2211640
